datafusion/physical-expr/src/array_expressions.rs

```diff
@@ -1358,6 +1360,94 @@ macro_rules! to_string {
     }};
 }
 
+fn union_generic_lists<OffsetSize: OffsetSizeTrait>(
+    l: &GenericListArray<OffsetSize>,
+    r: &GenericListArray<OffsetSize>,
+) -> Result<GenericListArray<OffsetSize>, DataFusionError> {
+    let converter =
+        RowConverter::new(vec![SortField::new(l.value_type().clone())]).unwrap();
+    
+    let nulls = NullBuffer::union(l.nulls(), r.nulls());
+    let field = Arc::new(Field::new(
+        "item",
+        l.value_type().to_owned(),
+        l.is_nullable(),
```

Comment:
Note that your current implementation cant pass column-wise test cases.

Example test case
```text
statement ok
CREATE TABLE arrays_with_repeating_elements
AS VALUES
  ([1], [2]),
  ([2, 3], [3]),
  ([3], [3, 4])
;

query ?
select array_union(column1, column2) from arrays_with_repeating_elements;
----
[1, 2]
[2, 3]
[3, 4]

statement ok
drop table arrays_with_repeating_elements;
```

The reason why your test fails is because 
```text
Invalid argument error: column types must match schema types, expected List\(Field \{ name: "item", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: \{\} \}\) but found List\(Field \{ name: "item", data_type: Int64, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: \{\} \}\) at column index 0
```

`l.is_nullable()` count number of nulls which is not always the same as the `nullable` in List field. In this case it return false while the expected value is true. The return type of `array_union` should be the same as the return type in `BuiltinScalarFunction`.  You can directly pass the `&FieldRef` into `union_generic_lists` since the field before and after `array_union` should be the same.

Reply:
I am looking at the ArrayRef and Lists API, I don't find how I could retrieve the FieldRef I am looking for, where is it?

Reply:
Fixed with https://github.com/apache/arrow-datafusion/pull/7897/commits/648ae5d659142082d085418dc7b84ad857ba8136